runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -38,13 +38,401 @@ TR::TreeLowering::optDetailString() const throw()
+   prevBlock->append(TR::TreeTop::create(comp, storeNode));
+
+   // insert acmpeq for fastpath, taking care to set the proper register dependencies
+   // Any register dependencies added by splitPostGRA will now be on the BBExit for
+   // the call block.  As the ifacmpeq branching around the call block will reach the same
+   // target block, copy any GlRegDeps from the end of the call block to the ifacmpeq
+   auto* ifacmpeqNode = TR::Node::createif(TR::ifacmpeq, anchoredCallArg1TT->getNode()->getFirstChild(), anchoredCallArg2TT->getNode()->getFirstChild(), targetBlock->getEntry());
+   if (callBlock->getExit()->getNode()->getNumChildren() > 0)
+      {
+      TR::Node* glRegDeps = TR::Node::create(TR::GlRegDeps);
+      TR::Node* depNode = NULL;
+
+      if (anchoredNode->getOpCodeValue() == TR::iRegLoad)
+         {
+         depNode = TR::Node::create(TR::PassThrough, 1, storeNode->getChild(0));
+         depNode->setGlobalRegisterNumber(storeNode->getGlobalRegisterNumber());
+         glRegDeps->addChildren(&depNode, 1);
+         }
+
+      ifacmpeqNode->addChildren(&glRegDeps, 1);
+
+      TR::Node* expectedDeps = callBlock->getExit()->getNode()->getFirstChild();
+      for (int i = 0; i < expectedDeps->getNumChildren(); ++i)
+         {
+         TR::Node* temp = expectedDeps->getChild(i);
+         if (depNode && temp->getGlobalRegisterNumber() == depNode->getGlobalRegisterNumber())
+            continue;
+         else if (temp->getOpCodeValue() == TR::PassThrough)
+            {
+            // PassThrough nodes cannot be commoned because doing so does not
+            // actually anchor the child, causing it's lifetime to not be extended
+            TR::Node* original = temp;
+            temp = TR::Node::create(original, TR::PassThrough, 1, original->getFirstChild());
+            temp->setLowGlobalRegisterNumber(original->getLowGlobalRegisterNumber());
+            temp->setHighGlobalRegisterNumber(original->getHighGlobalRegisterNumber());
+            }
+         glRegDeps->addChildren(&temp, 1);
+         }
+      }
+
+   prevBlock->append(TR::TreeTop::create(comp, ifacmpeqNode));
+   }
+
+/*
+ * If value types are enabled, and the value that is being assigned to the array
+ * element might be a null reference, lower the ArrayStoreCHK by splitting the
+ * block before the ArrayStoreCHK, and inserting a NULLCHK guarded by a check
+ * of whether the array's component type is a value type.
+ */
+void
+TR::TreeLowering::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+   {
+   traceMsg(comp(), "INVOKED!!!\n");
+   // Pattern match the ArrayStoreCHK operands to get the source of the assignment
+   // (sourceChild) and the array to which an element will have a value assigned (destChild)
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR::Node *sourceChild = firstChild->getSecondChild();
+   TR::Node *destChild = firstChild->getChild(2);
```

Comment:
Maybe this above code is assuming that the tree under the array store check will be a `wrtbari` whereas `astorei` is also a possibility ? i.e. I wonder if this will work for an `astorei`, e.g. is there a 3rd child ?

Reply:
Thanks, Vijay - I'll try out that case.  As this part of Leo's change is simply moving code from `J9CodeGenerator.cpp` to `TreeLowering.cpp` without any changes, I suggest that we not let this comment hold up the addition of the new optimization.  As I originally wrote this method, I will make any required changes to handle `astorei` correctly in a follow on pull request.

Reply:
Yes I am okay with that since you plan to address separately.

Reply:
Vijay @vijaysun-omr, I spent some time today trying to figure out how to get an `astorei` as a child of `ArrayStoreCHK`.  Do you know whether it is actually possible for an `ArrayStoreCHK` to have an `astorei` as a child, and if so how?  Or had I simply made a mistake in my [original comments](https://github.com/Leonardo2718/openj9/blob/6495f850594ff714c50eafd542aeb673b3fd191a/runtime/compiler/optimizer/TreeLowering.cpp#L344-L353) in this code?

Reply:
You might be right with respect to only the original comments in the PR being at issue. I don't think we allow anything but a write barrier under an array store check as is evidenced by a transformation that "eliminates" the write barrier https://github.com/eclipse/omr/blob/67ddc287341ab2bce7dfee85f1fde1f1e9b3889d/compiler/optimizer/VPHandlers.cpp#L3213

An `astorei` node may be possible when both the array store check and the write barrier are proven to not be required (i.e. the array store check is removed and the write barrier node could be changed to be a store) but this is not the case that is relevant for this PR. 

So you may only have to add in an assert to make sure it is a write barrier as we expect and fix up your comment in the PR :)

Reply:
Adjusted the comments and added an assertion in pull request #12126.  Vijay @vijaysun-omr, may I ask you to take a look?